### ISSUE TITLE
Authentication failure in protocol adapter with hono example

### ISSUE THREAD

**Issue by calohmn** [eclipse/hono#209]
Using the `docker-compose.yml` from the Hono Getting Started guide causes authentication failure errors in the protocol adapters.

PR #208 addresses issues concerning wrong credentials and wrong auth host name.

However, after applying these changes, there is still a `Failed to authenticate` exception on protocol adapter startup. The reason isn't directly obvious to me. Log output of the Auth container appears to show no issue. In the Hono log there is this output:
````
09:29:22.801 [vert.x-eventloop-thread-1] DEBUG o.e.h.s.auth.HonoSaslAuthenticator - authentication failed: Timed out after waiting 1000(ms) for a reply. address: 1
09:29:22.863 [vert.x-eventloop-thread-1] DEBUG o.e.h.s.auth.HonoSaslAuthenticator - authentication failed: Timed out after waiting 1000(ms) for a reply. address: 2
````

**Comment by sophokles73** [comment]
The message in the log is the (failed) result of an attempt to authenticate a client (the protocol adapters I assume) by delegating to the auth server. This is probably due to the fact that the auth server is not (yet) up and running. However, because the protocol adapters try again and again to connect to the Hono server, the authentication request eventually succeeds and everything is fine. The failed connection attempts do not show up in the adapters' log by default.

I assume that the protocol adapters do work, don't they?

**Comment by sophokles73** [comment]
I have taken another look and I believe that the client (the protocol adapter in this case) currently is not able to handle this situation because the `HonoClient` does not re-try on an authentication failure (which makes kind of sense because the credentials will not suddenly be correct on the next attempt). I think we will need to prevent this from happening by means of making the Hono server open its ports only once the services it depends on become available, i.e. the Dispatch Router and the authentication server.

**Comment by calohmn** [comment]
Yes, the authentication failure in the protocol adapter was indeed due to the auth server not being fully started yet.

Concerning the re-try: With the new setup of having an external Auth server, as it is currently implemented, an authentication failure doesn't necessarily mean that credentials are wrong - it can just be that the auth server is currently down.
Would it be possible to return a different error in that case (something like "service anavailable"), so that such an error can be handled with a re-try in the protocol adapter? 
If not, I would opt for an unconditional re-try or, as the adapter is in kind of dead state then, to shutdown the adapter.

**Comment by sophokles73** [comment]
> If not, I would opt for an unconditional re-try or, as the adapter is in kind of dead state then, to shutdown the adapter.

This is what I prefer as well. In conjunction with a *health check* HTTP resource provided by protocol adapters it should be possible for e.g. kubernetes to only expose working protocol adapter instances to external clients. So this is the approach I am following now ...

**Comment by calohmn** [comment]
After having applied the workaround of restarting the adapters once the auth-server was fully started, the example scenario worked. Device registration was successful and telemetry data was sent and consumed successfully.

However, after a certain time, the adapters returned errors: `Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}`

Here is the relevant log output of the **MQTT adapter**:
````
09:08:02.169 [vert.x-eventloop-thread-0] INFO o.e.h.a.m.VertxBasedMqttProtocolAdapter - Connection request from client esp8266.600194195a1b
09:08:02.212 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - reusing existing registration client for [DEFAULT_TENANT]
09:08:02.212 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - reusing existing message sender [target: telemetry/DEFAULT_TENANT, credit: 212]
09:08:02.238 [vert.x-eventloop-thread-0] DEBUG o.e.h.c.i.AbstractRequestResponseClient - received response [correlation ID: reg-client-171, status: 200]

09:08:08.454 [vert.x-eventloop-thread-0] INFO o.e.h.a.m.VertxBasedMqttProtocolAdapter - Connection request from client esp8266.5ccf7f9509f1
09:08:08.518 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - reusing existing registration client for [DEFAULT_TENANT]
09:08:08.519 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - reusing existing message sender [target: telemetry/DEFAULT_TENANT, credit: 211]
09:08:08.545 [vert.x-eventloop-thread-0] DEBUG o.e.h.c.i.AbstractRequestResponseClient - received response [correlation ID: reg-client-172, status: 200]

09:08:08.550 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.AbstractSender - sender [telemetry/DEFAULT_TENANT] closed: Error{condition=hono:no-downstream-consumer, description='no downstream consumer available for data', info=null}

09:08:12.266 [vert.x-eventloop-thread-0] INFO o.e.h.a.m.VertxBasedMqttProtocolAdapter - Connection request from client esp8266.600194195a1b
09:08:12.309 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - reusing existing registration client for [DEFAULT_TENANT]
09:08:12.310 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - creating new message sender for telemetry/DEFAULT_TENANT
09:08:12.313 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.AbstractSender - opening sender [telemetry/DEFAULT_TENANT] failed: Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}
09:08:12.314 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - failed to create new message sender for telemetry/DEFAULT_TENANT
io.vertx.core.impl.NoStackTraceThrowable: Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}
09:08:12.314 [vert.x-eventloop-thread-0] DEBUG o.e.h.a.m.VertxBasedMqttProtocolAdapter - cannot process message from [esp8266.600194195a1b]
io.vertx.core.impl.NoStackTraceThrowable: Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}
09:08:12.314 [vert.x-eventloop-thread-0] DEBUG o.e.h.a.m.VertxBasedMqttProtocolAdapter - closing connection with client [client ID: esp8266.600194195a1b]
09:08:12.314 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.AbstractSender - sender [telemetry/DEFAULT_TENANT] closed: Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}

09:08:12.336 [vert.x-eventloop-thread-0] DEBUG o.e.h.c.i.AbstractRequestResponseClient - received response [correlation ID: reg-client-173, status: 200]
09:08:18.531 [vert.x-eventloop-thread-0] INFO o.e.h.a.m.VertxBasedMqttProtocolAdapter - Connection request from client esp8266.5ccf7f9509f1
09:08:18.585 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - reusing existing registration client for [DEFAULT_TENANT]
09:08:18.585 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - creating new message sender for telemetry/DEFAULT_TENANT
09:08:18.589 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.AbstractSender - opening sender [telemetry/DEFAULT_TENANT] failed: Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}
09:08:18.589 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.HonoClientImpl - failed to create new message sender for telemetry/DEFAULT_TENANT
io.vertx.core.impl.NoStackTraceThrowable: Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}
09:08:18.589 [vert.x-eventloop-thread-0] DEBUG o.e.h.a.m.VertxBasedMqttProtocolAdapter - cannot process message from [esp8266.5ccf7f9509f1]
io.vertx.core.impl.NoStackTraceThrowable: Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}
09:08:18.589 [vert.x-eventloop-thread-0] DEBUG o.e.h.a.m.VertxBasedMqttProtocolAdapter - closing connection with client [client ID: esp8266.5ccf7f9509f1]
09:08:18.590 [vert.x-eventloop-thread-0] DEBUG o.e.hono.client.impl.AbstractSender - sender [telemetry/DEFAULT_TENANT] closed: Error{condition=amqp:unauthorized-access, description='unauthorized', info=null}
````

In the **Hono** log, there is only the message `subject [mqtt-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]`.
Log output directly before the first error.
````
09:05:00.961 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono MQTT Adapter-c6e41c84-1b38-46ef-b943-572630e28dd9] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:05:00.961 [vert.x-eventloop-thread-1] DEBUG o.eclipse.hono.auth.AuthoritiesImpl - no claim for key [r:telemetry/DEFAULT_TENANT]
09:05:00.961 [vert.x-eventloop-thread-1] DEBUG o.eclipse.hono.auth.AuthoritiesImpl - found claim [key: r:telemetry/*, activities: W] matching intent [WRITE]
09:05:00.973 [vert.x-eventloop-thread-1] DEBUG o.e.hono.server.SenderFactoryImpl - sender [telemetry/DEFAULT_TENANT] for container [Hono.Example.Router] open
09:05:00.973 [vert.x-eventloop-thread-1] INFO o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - created downstream sender [con: 668a3d45-d45d-4790-ac4a-3ce86ee58c6d, link: 0b7c7e84-01f3-4f8a-9dc8-7b24146c02fd]
09:05:00.973 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.impl.TelemetryEndpoint - accepted link from telemetry client [0b7c7e84-01f3-4f8a-9dc8-7b24146c02fd]
09:05:00.973 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - received FLOW from downstream sender [con:668a3d45-d45d-4790-ac4a-3ce86ee58c6d, link: 0b7c7e84-01f3-4f8a-9dc8-7b24146c02fd, credits: 250, queued: 0, drain: false
09:05:00.973 [vert.x-eventloop-thread-1] DEBUG o.e.h.s.amqp.UpstreamReceiverImpl - replenishing client [0b7c7e84-01f3-4f8a-9dc8-7b24146c02fd] with 250 credits
09:05:00.989 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - received FLOW from downstream sender [con:668a3d45-d45d-4790-ac4a-3ce86ee58c6d, link: 0b7c7e84-01f3-4f8a-9dc8-7b24146c02fd, credits: 250, queued: 0, drain: false
09:05:01.232 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - received FLOW from downstream sender [con:668a3d45-d45d-4790-ac4a-3ce86ee58c6d, link: 0b7c7e84-01f3-4f8a-9dc8-7b24146c02fd, credits: 250, queued: 0, drain: false
09:05:02.210 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono REST Adapter-070fd1e4-18df-412b-946b-08d52e6816bb] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:05:02.210 [vert.x-eventloop-thread-1] DEBUG o.eclipse.hono.auth.AuthoritiesImpl - no claim for key [r:telemetry/DEFAULT_TENANT]
09:05:02.213 [vert.x-eventloop-thread-1] DEBUG o.eclipse.hono.auth.AuthoritiesImpl - found claim [key: r:telemetry/*, activities: W] matching intent [WRITE]
09:05:02.226 [vert.x-eventloop-thread-1] DEBUG o.e.hono.server.SenderFactoryImpl - sender [telemetry/DEFAULT_TENANT] for container [Hono.Example.Router] open
09:05:02.226 [vert.x-eventloop-thread-1] INFO o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - created downstream sender [con: a182743c-d15a-4ae8-8e57-ceb7d2bf186b, link: 5ff1b063-ea40-4ea5-8a4e-824a30b72089]
09:05:02.226 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.impl.TelemetryEndpoint - accepted link from telemetry client [5ff1b063-ea40-4ea5-8a4e-824a30b72089]
09:05:02.226 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - received FLOW from downstream sender [con:a182743c-d15a-4ae8-8e57-ceb7d2bf186b, link: 5ff1b063-ea40-4ea5-8a4e-824a30b72089, credits: 250, queued: 0, drain: false
09:05:02.227 [vert.x-eventloop-thread-1] DEBUG o.e.h.s.amqp.UpstreamReceiverImpl - replenishing client [5ff1b063-ea40-4ea5-8a4e-824a30b72089] with 250 credits
09:05:02.243 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - received FLOW from downstream sender [con:a182743c-d15a-4ae8-8e57-ceb7d2bf186b, link: 5ff1b063-ea40-4ea5-8a4e-824a30b72089, credits: 250, queued: 0, drain: false
...
09:06:52.963 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - received FLOW from downstream sender [con:a182743c-d15a-4ae8-8e57-ceb7d2bf186b, link: 5ff1b063-ea40-4ea5-8a4e-824a30b72089, credits: 250, queued: 0, drain: false
09:07:00.733 [vert.x-eventloop-thread-1] WARN o.e.h.e.i.ForwardingEventDownstreamAdapter - lost connection to downstream container [Hono.Example.Router], closing upstream receivers ...
09:07:00.933 [vert.x-eventloop-thread-1] INFO o.e.h.e.i.ForwardingEventDownstreamAdapter - attempting to re-connect to downstream container
09:07:00.934 [vert.x-eventloop-thread-1] INFO o.e.h.s.DownstreamClientConfigProperties - using certificates from file [/run/secrets/trusted-certs.pem] as trust anchor
09:07:00.934 [vert.x-eventloop-thread-1] INFO o.e.h.s.DownstreamClientConfigProperties - using key & certificate from PKCS12 key store [/run/secrets/honoKeyStore.p12] for identity
09:07:00.934 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qdrouter:5673]
09:07:01.718 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://qdrouter:5673], opening connection ...
09:07:01.720 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connection to container [Hono.Example.Router] at [amqps://qdrouter:5673] open
09:07:01.808 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - received FLOW from downstream sender [con:668a3d45-d45d-4790-ac4a-3ce86ee58c6d, link: 0b7c7e84-01f3-4f8a-9dc8-7b24146c02fd, credits: 249, queued: 0, drain: false
...
09:08:03.502 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - received FLOW from downstream sender [con:a182743c-d15a-4ae8-8e57-ceb7d2bf186b, link: 5ff1b063-ea40-4ea5-8a4e-824a30b72089, credits: 250, queued: 0, drain: false
09:08:08.545 [vert.x-eventloop-thread-1] WARN o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - lost connection to downstream container [Hono.Example.Router], closing upstream receivers ...
09:08:08.547 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.impl.TelemetryEndpoint - closing receiver for client [5ff1b063-ea40-4ea5-8a4e-824a30b72089]
09:08:08.548 [vert.x-eventloop-thread-1] DEBUG o.e.h.t.impl.TelemetryEndpoint - closing receiver for client [0b7c7e84-01f3-4f8a-9dc8-7b24146c02fd]
09:08:08.746 [vert.x-eventloop-thread-1] INFO o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - attempting to re-connect to downstream container
09:08:08.746 [vert.x-eventloop-thread-1] INFO o.e.h.s.DownstreamClientConfigProperties - using certificates from file [/run/secrets/trusted-certs.pem] as trust anchor
09:08:08.746 [vert.x-eventloop-thread-1] INFO o.e.h.s.DownstreamClientConfigProperties - using key & certificate from PKCS12 key store [/run/secrets/honoKeyStore.p12] for identity
09:08:08.746 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qdrouter:5673]
09:08:08.909 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://qdrouter:5673], opening connection ...
09:08:08.910 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connection to container [Hono.Example.Router] at [amqps://qdrouter:5673] open
````
Errors start here:
````
09:08:09.556 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono REST Adapter-070fd1e4-18df-412b-946b-08d52e6816bb] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:08:09.557 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - subject [rest-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]
09:08:12.309 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono MQTT Adapter-c6e41c84-1b38-46ef-b943-572630e28dd9] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:08:12.309 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - subject [mqtt-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]
09:08:12.983 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono REST Adapter-070fd1e4-18df-412b-946b-08d52e6816bb] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:08:12.983 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - subject [rest-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]
09:08:18.584 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono MQTT Adapter-c6e41c84-1b38-46ef-b943-572630e28dd9] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:08:18.584 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - subject [mqtt-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]
09:08:19.004 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono REST Adapter-070fd1e4-18df-412b-946b-08d52e6816bb] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:08:19.004 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - subject [rest-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]
09:08:22.408 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono MQTT Adapter-c6e41c84-1b38-46ef-b943-572630e28dd9] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:08:22.408 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - subject [mqtt-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]
09:08:23.102 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono REST Adapter-070fd1e4-18df-412b-946b-08d52e6816bb] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:08:23.102 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - subject [rest-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]
09:08:28.660 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - client [Hono MQTT Adapter-c6e41c84-1b38-46ef-b943-572630e28dd9] wants to open a link for sending messages [address: Target{address='telemetry/DEFAULT_TENANT', durable=NONE, expiryPolicy=SESSION_END, timeout=0, dynamic=false, dynamicNodeProperties=null, capabilities=null}]
09:08:28.660 [vert.x-eventloop-thread-1] DEBUG org.eclipse.hono.server.HonoServer - subject [mqtt-adapter@HONO] is not authorized to WRITE to [telemetry/DEFAULT_TENANT]
````

In the **auth-server** log, there is no output at all in that time period and later on. 

I've restarted the auth-server after that. But that didn't solve the problem. In the auth-server logs I saw no incoming requests at all.

**Comment by sophokles73** [comment]
Yes, I guess this is due to the JWT tokens expiring after around 10 minutes (does that match the time after which the errors show up in your case?).

We currently have no way of refreshing the tokens automatically. This requires some changes in the vertx-proton API (which I will do over the next days/weeks). For the time being, we will simply close the connection with the client (adapter) so that it re-establishes a connection and gets a fresh token.

**Comment by calohmn** [comment]
Yes, it likely was such a 10min time frame after which errors occurred.

**Comment by sophokles73** [comment]
should be fixed in a70be270a72f71bc5ad60b785d0a1cc09f70c19d

@calohmn can you take it for a ride and see if this fixes the issue for you?

**Comment by calohmn** [comment]
It looks good now.
